### DISCUSSION TITLE
useShallow on deep dictionary

### DISCUSSION THREAD

**Discussion by pthieu** [pmndrs/zustand#2094] [General] [answered]
I have a dictionary that is something like this:
```
const useCaptionsEditStore = createWithEqualityFn<CaptionsEditStore>()({
  dict: {
    id1: {...} 
    id2: {///}
  }
})
```

I get the whole dict like
```
const {dict} = useStore(...)
```

Unfortunately, when i update any specific item in `dict`, it will re-render any component that has imported the whole dict. With `useShallow`, it looked like my problems would be solved but it seems like whatever you're comparing, it will return, so:

```
...
useShallow((state)=>Object.keys(state.dict))
...
```

Is there a way to compare the keys of the dict, but return the actual content inside?

My current solution is to implement a dictionary slice getting fn where I pass in `id`'s, but would be great if I didn't have to do this

**Comment by dbritto-dev** [comment, marked as answer]
@pthieu you can build your own `useShallow` 

```ts
export function useShallow<S, U>(selector: (state: S) => U): (state: S) => U {
  const prev = useRef<U>()

  return (state) => {
    const next = selector(state)
    return shallow(prev.current, next)
      ? (prev.current as U)
      : (prev.current = state) // we are returning state instead of next due to was created by selector
  }
}
```

**Reply by pthieu** [reply]
ohhh I didn't know you could do this
thanks, let me try

can this be placed at the store level?

**Reply by dbritto-dev** [reply]
@pthieu yeah, you can colocate that in a different file if you want. Keep me in the loop.

**Reply by pthieu** [reply]
@dbritto-dev oh I mean if I can do something to the equivalent of this:

```
createWithEqualityFn(..., shallow)
```

so every time I do
```
const { dict }  = useStore(...)
```

It will always use the custom `useShallow` for that dict?

**Reply by dbritto-dev** [reply]
@pthieu yeah, now the preferred way is using `create` and `useShallow` instead of `createWithEqualityFn` so you don't need to import 'zustand/traditional'

**Reply by dbritto-dev** [reply]
@pthieu keep me in the loop

**Comment by dai-shi** [comment]
> Unfortunately, when i update any specific item in dict, it will re-render any component that has imported the whole dict.

Do you mean you update a property of an item in dict? In that case, you would need "deep" comparison.

```js
createWithEqualityFn(..., shallow)
```
If that works, then shallow comparison would work fine.

**Reply by dbritto-dev** [reply]
He want to use `useShallow` and look for an slice but return the whole state, for instance compare `id1` but return `id1` and `id2`

**Reply by pthieu** [reply]
@dai-shi yeah, I just want to prevent re-renders for components that are importing/subscribing to the entire dictionary when *another* component makes a change to a field in one of the dictionary's elements

right now i have a getter function to get the elements from using `id[]` as an input so I don't subscribe to the dictionary updates

the custom `useShallow` will work, but I haven't had time to figure out how to apply it to the `store` so I don't have to define it every time I import/subscribe to the entire dictionary

**Reply by dai-shi** [reply]
Thanks for the clarification.

Here's what I would do:
```js
const useCaptionsEditStore = createWithEqualityFn((set) => ({
  dict: {
    id1: { /* ... */ },
    id2: { /* ... */ },
  },
})); // the latest version doesn't require the second argument.

  const dict = useCaptionsEditStore((state) => state.dict, (a, b) => a.id1 === b.id1);
```

Of course, creating a custom hook like `useShallow` works too.

**Reply by pthieu** [reply]
@dai-shi thanks, is there a way to do it in the caption store itself?
```
const useCaptionsEditStore = createWithEqualityFn((set) => ({
    dict: {
      id1: { /* ... */ },
      id2: { /* ... */ },
    },
  })
  // something like right here we can specify for `dict`, the comparator should be `(a, b) => a.id1 === b.id1`
); 
```

**Reply by dbritto-dev** [reply]
A custom hook like `useShallow` best suits your case

**Reply by dai-shi** [reply]
> do it in the caption store itself?

In that case, `createWithEqualityFn` might be better, but your custom comparator for the second argument will be used for all selectors.
`(a, b) => a.id1 === b.id1` doesn't work if you do `useCaptionsEditStore((state) => state.somethingElse)`.

So, what makes sense is to export only limited store hook.

```js
const useCaptionsEditStore_INTERNAL = createWithEqualityFn((set) => ({
  dict: {
    id1: { /* ... */ },
    id2: { /* ... */ },
  },
})); 
export const useCaptionsEditStore = () => useCaptionsEditStore_INTERNAL((state) => state.dict, (a, b) => a.id1 === b.id1);
```
